### PR TITLE
Refactor/popover

### PR DESCRIPTION
**Pull request by Tibing**
### Please read and mark the following check list before creating a pull request:

 - [x] I read and followed the [CONTRIBUTING.md](https://github.com/akveo/nebular/blob/master/CONTRIBUTING.md) guide.
 - [x] I read and followed the [New Feature Checklist](https://github.com/akveo/nebular/blob/master/DEV_DOCS.md#new-feature-checklist) guide.
 
 #### Short description of what this resolves:
Give the user capability pass context to the popover.

### CODE REVIEW

File: src/framework/theme/components/popover/popover.component.ts [hunk 1]
```diff
@@ -46,6 +49,12 @@ export class NbPopoverComponent {
   @Input()
   content: NbPopoverContent;
 
+  /**
+   * Context which will be passed to rendered component instance.
+   * */
+  @Input()
+  context: Object;
```
Comment (nnixaa) [inline src/framework/theme/components/popover/popover.component.ts]: Shouldn't we validate the `context`? For example, clear out `functions` from it?

Comment (nnixaa) [reply src/framework/theme/components/popover/popover.component.ts]: It's just a thought, not sure we really need this.

File: src/framework/theme/components/popover/popover.component.ts [hunk 2]
```diff
@@ -61,6 +70,24 @@ export class NbPopoverComponent {
...
   positionLeft: number;
 
+  /**
+   * If content type is TemplateRef we're passing context as template outlet param.
+   * But if we have custom component content we're just assigning passed context to the component instance.
+   * */
+  @ViewChild(NgComponentOutlet)
+  set componentOutlet(el) {
+    if (this.isComponent) {
+      Object.assign(el._componentRef.instance, this.context);
```
Comment (nnixaa) [inline src/framework/theme/components/popover/popover.component.ts]: is this the best way to access the component reference?

File: src/framework/theme/components/popover/popover.directive.ts [hunk 3]
```diff
@@ -212,6 +218,7 @@ export class NbPopoverDirective implements OnInit, OnDestroy {
       .subscribe((containerRef: ComponentRef<NbPopoverComponent>) => {
         this.containerRef = containerRef;
         this.patchPopoverContent(this.content);
+        this.patchPopoverContext(this.context);
```
Comment (nnixaa) [inline src/framework/theme/components/popover/popover.directive.ts]: do we need two functions here? can be just `patchPopover(this.content, this.context)`

File: src/app/layout-test/theme-dynamic-test.component.ts [hunk 4]
```diff
@@ -12,11 +12,13 @@ import { NbThemeService } from '@nebular/theme';
...
   template: `
     <div>
-      <strong>hello from dynamically inserted component</strong>
+      <strong>hello from dynamically inserted component: {{text}}</strong>
     </div>
   `,
 })
-export class NbDynamicToAddComponent {}
+export class NbDynamicToAddComponent {
+  text: string = '';
```
Comment (nnixaa) [inline src/app/layout-test/theme-dynamic-test.component.ts]: we definitely need to test this

File: src/framework/theme/components/popover/popover.component.ts (hunk 1 above)
Comment (Tibing) [reply src/framework/theme/components/popover/popover.component.ts]: I don't think it's important. Maybe the user will want to pass functions to the content component?

File: src/framework/theme/components/popover/popover.component.ts (hunk 2 above)
Comment (Tibing) [reply src/framework/theme/components/popover/popover.component.ts]: Yep, because we don't have component type and can query only by *NgComponentOutlet*

Comment (nnixaa) [reply src/framework/theme/components/popover/popover.component.ts]: at least we can do `@ViewChild(NgComponentOutlet, {read: ViewContainerRef})`

Comment (Tibing) [reply src/framework/theme/components/popover/popover.component.ts]: Ok, I'll check it

File: src/framework/theme/components/popover/popover.directive.ts (hunk 3 above)
Comment (Tibing) [reply src/framework/theme/components/popover/popover.directive.ts]: fixed

File: src/app/layout-test/theme-dynamic-test.component.ts (hunk 4 above)
Comment (Tibing) [reply src/app/layout-test/theme-dynamic-test.component.ts]: tested

**Pull request merged**